core/src/main/java/com/datastrato/gravitino/lock/LockManager.java

```diff
@@ -0,0 +1,172 @@
+ * the following
+ *
+ * <pre>
+ *   /                                    readLock
+ *   /metalake                            readLock
+ *   /metalake/catalog                    readLock
+ *   /metalake/catalog/db1                readLock
+ *   /metalake/catalog/db/table1          writeLock
+ * </pre>
+ *
+ * If the lock manager fails to lock the resource path, it will release all the locks that have been
+ * locked in the inverse sequences it locks the resource path.
+ */
+public class LockManager {
+  private static final Logger LOG = LoggerFactory.getLogger(LockManager.class);
+  private static final long TIME_AFTER_LAST_ACCESS_TO_EVICT_IN_MS = 5 * 60 * 1000L; // 1 hour
+
+  @VisibleForTesting final TreeLockNode treeLockRootNode = new TreeLockNode(NameIdentifier.ROOT);
+  private final ThreadLocal<Stack<TreeLockPair>> holdingLocks = ThreadLocal.withInitial(Stack::new);
+  private final ScheduledThreadPoolExecutor lockCleaner;
+
+  static class TreeLockPair {
+    private final TreeLockNode treeLockNode;
+    private final LockType lockType;
+
+    public TreeLockPair(TreeLockNode treeLockNode, LockType lockType) {
+      this.treeLockNode = treeLockNode;
+      this.lockType = lockType;
+    }
+
+    public static TreeLockPair of(TreeLockNode treeLockNode, LockType lockType) {
+      return new TreeLockPair(treeLockNode, lockType);
+    }
+  }
+
+  public LockManager() {
+    this.lockCleaner =
+        new ScheduledThreadPoolExecutor(
+            1,
+            new ThreadFactoryBuilder()
+                .setDaemon(true)
+                .setNameFormat("tree-lock-cleaner-%d")
+                .build());
+
+    lockCleaner.scheduleAtFixedRate(
+        () ->
+            treeLockRootNode
+                .getAllChildren()
+                .forEach(
+                    child ->
+                        evictStaleNodes(
+                            TIME_AFTER_LAST_ACCESS_TO_EVICT_IN_MS, child, treeLockRootNode)),
+        1,
+        1,
+        TimeUnit.MINUTES);
+  }
+
+  /**
+   * Evict the stale nodes from the tree lock node if the last access time is earlier than the given
+   * time.
+   *
+   * @param timeAfterLastAccessToEvictInMs The time after last access to evict the node.
+   * @param treeNode The tree lock node to evict.
+   * @param parent The parent of the tree lock node.
+   */
+  @VisibleForTesting
+  void evictStaleNodes(
+      long timeAfterLastAccessToEvictInMs, TreeLockNode treeNode, TreeLockNode parent) {
+    if (treeNode.getLastAccessTime()
+        < System.currentTimeMillis() - timeAfterLastAccessToEvictInMs) {
+      // We need to assure that during the getLastAccessTime and removeChild, no other thread
+      // can access and change the tree node.
+      boolean success = treeNode.tryLock(LockType.WRITE);
+      if (!success) {
+        // This node is locked by another thread, we will skip it and try to evict its children.
+        treeNode
```

Comment:
If this thread obtains an expired lock, but the runtime thread is holding the lock without locking it at this time, the lock object will be dropped from the lock tree. Consequently, the runtime thread will end up using an unmanaged lock object.
I suggest you use the reference count to resolve the problems

Reply:
I see, let me think about it.